helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java

```diff
@@ -646,10 +652,14 @@ private void subscribeForChanges(NotificationContext.Type callbackType, String p
               break;
             }
             default: {
-              List<String> childNames = _zkClient.getChildren(path);
-              if (childNames != null) {
-                for (String childName : childNames) {
-                  String childPath = path + "/" + childName;
+              // When callback type is FINALIZE, subscribeChildChange doesn't
+              // get children list. We need to read children to unsubscribe data change listeners.
+              if (callbackType == Type.FINALIZE) {
+                children = _zkClient.getChildren(path);
+              }
```

Comment:
As we discussed, I think this check is not necessary. The later check for children != null is good enough.
If the node was not created and then create during this callback processing, then we will have another callback to process the event since the watcher has been added.

Reply:
Moved the logic to subscribeChildChange: for FINALIZE, it also returns the children.